### PR TITLE
WIP rbd/cache: fix UT failures when RWL is enabled

### PR DESCRIPTION
**Pull request by lixiaoy1**
Signed-off-by: Li, Xiaoyan <xiaoyan.li@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/librbd/internal.cc [hunk 1]
```diff
@@ -727,7 +727,8 @@ int validate_pool(IoCtx &io_ctx, CephContext *cct) {
     uint64_t order = *c_order;
 
     ImageOptions opts;
-    opts.set(RBD_IMAGE_OPTION_FEATURES, features);
+    uint64_t new_features = features & ~RBD_FEATURE_DIRTY_CACHE;
```
Comment (dillaman) [inline src/librbd/internal.cc:730]: Add `RBD_FEATURE_DIRTY_CACHE` to `RBD_FEATURES_IMPLICIT_ENABLE`

File: src/osdc/Striper.cc [hunk 2]
```diff
@@ -390,12 +390,14 @@ void Striper::StripedReadResult::add_partial_result(
 {
   ldout(cct, 10) << "add_partial_result(" << this << ") " << bl.length()
 		 << " to " << buffer_extents << dendl;
+  uint64_t off = 0;
   for (auto p = buffer_extents.cbegin(); p != buffer_extents.cend(); ++p) {
-    pair<bufferlist, uint64_t>& r = partial[p->first];
+    pair<bufferlist, uint64_t>& r = partial[off];
```
Comment (dillaman) [inline src/osdc/Striper.cc:395]: This change is wrong -- if something is broken, it's implying an issue in librbd/RWL

File: src/librbd/internal.cc (hunk 1 above)
Comment (lixiaoy1) [reply src/librbd/internal.cc:730]: Currently DIRTY_CACHE is already in RBD_FEATURES_IMPLICIT_ENABLE. 
In fact, I think the best place to handle RBD_FEATURES_IMPLICIT_ENABLE features is in CloneRequest.cc. 
I am concerned why we need the attribute m_use_p_features 
https://github.com/ceph/ceph/blob/master/src/librbd/image/CloneRequest.cc#L59  
If calling with features, m_use_p_features  is false.
```cpp
  if (m_opts.get(RBD_IMAGE_OPTION_FEATURES, &m_features) == 0) {
    if (m_features & ~RBD_FEATURES_ALL) {
      lderr(m_cct) << "librbd does not support requested features" << dendl;
      complete(-ENOSYS);
      return;
    }
    m_use_p_features = false;
  }
```
It blocks to exclude implicit features: 
```c
 if (m_use_p_features) {
    m_features = (p_features & ~RBD_FEATURES_IMPLICIT_ENABLE);
  }
``` 
https://github.com/ceph/ceph/blob/master/src/librbd/image/CloneRequest.cc#L216-L218

File: src/osdc/Striper.cc (hunk 2 above)
Comment (lixiaoy1) [reply src/osdc/Striper.cc]: Yes.

File: src/osdc/Striper.cc [hunk 3]
```diff
@@ -524,13 +524,14 @@ uint64_t Striper::StripedReadResult::assemble_result(
     bufferlist *bl)
 {
   ldout(cct, 10) << "assemble_result(" << this << ")" << dendl;
+  uint64_t off = 0;
```
Comment (lixiaoy1) [inline src/osdc/Striper.cc:527]: @dillaman  How about this change in the function? 
```cpp
uint64_t Striper::StripedReadResult::assemble_result(
    CephContext *cct, std::map<uint64_t, uint64_t> *extent_map,
    bufferlist *bl)
```
Before updating, extent_map.first is image offset. Now it is the offset in the bl.

Comment (lixiaoy1) [inline src/osdc/Striper.cc:527]: Please ignore the updates. I think I need to remap it in https://github.com/ceph/ceph/blob/master/src/librbd/io/ReadResult.cc#L162

File: src/librbd/image/CloneRequest.cc [hunk 4]
```diff
@@ -99,7 +98,6 @@ void CloneRequest<I>::validate_options() {
       complete(-ENOSYS);
       return;
     }
-    m_use_p_features = false;
```
Comment (dillaman) [inline src/librbd/image/CloneRequest.cc:102]: Huh? If we explicitly set the new features to use, we shouldn't inherit features from the parent.

File: src/librbd/image/CloneRequest.cc [hunk 5]
```diff
@@ -213,9 +211,7 @@ void CloneRequest<I>::validate_parent() {
     close_parent();
     return;
   }
-  if (m_use_p_features) {
-    m_features = (p_features & ~RBD_FEATURES_IMPLICIT_ENABLE);
-  }
+  m_features = (p_features & ~RBD_FEATURES_IMPLICIT_ENABLE);
```
Comment (dillaman) [inline src/librbd/image/CloneRequest.cc:214]: I think this is what you really need:

```
if (m_use_p_features) {
  m_features = p_features;
}
m_features &= ~RBD_FEATURES_IMPLICIT_ENABLE;
```

Comment (lixiaoy1) [reply src/librbd/image/CloneRequest.cc:214]: Right.

Comment (lixiaoy1) [reply src/librbd/image/CloneRequest.cc:214]: I submit it in a standalone PR https://github.com/ceph/ceph/pull/37743.

**Comment by lixiaoy1** [comment]
Thank you @dillaman . I created the PR https://github.com/ceph/ceph/pull/37750 for extent_map and data miscompare issues when deep copy. This PR will be closed.

**Pull request closed without merging**